File: backend/src/models/Order.ts
import { Schema, model, type InferSchemaType } from 'mongoose';

const addressSchema = new Schema({
  addressLine: { type: String, required: true, trim: true },
  city: { type: String, required: true, trim: true },
  region: { type: String, trim: true },
  postalCode: { type: String, trim: true },
  location: {
    type: { type: String, enum: ['Point'], required: true, default: 'Point' },
    coordinates: { type: [Number], required: true, validate: { validator: (value: number[]) => value.length === 2 } }
  }
}, { _id: false });

const dimensionsSchema = new Schema({ length: { type: Number, min: 0 }, width: { type: Number, min: 0 }, height: { type: Number, min: 0 } }, { _id: false });
const priceBreakdownSchema = new Schema({ base: { type: Number, min: 0, required: true }, distance: { type: Number, min: 0, default: 0 }, surcharge: { type: Number, min: 0, default: 0 }, discount: { type: Number, min: 0, default: 0 }, total: { type: Number, min: 0, required: true }, currency: { type: String, default: 'USD' } }, { _id: false });
const recipientSchema = new Schema({ phone: { type: String, trim: true }, verificationPin: { type: String, trim: true, minlength: 4, maxlength: 8 } }, { _id: false });

const orderSchema = new Schema({
  trackingNumber: { type: String, required: true, unique: true, index: true, trim: true },
  merchantId: { type: Schema.Types.ObjectId, ref: 'MerchantProfile', required: true, index: true },
  assignedDriverId: { type: Schema.Types.ObjectId, ref: 'DriverProfile', index: true },
  pickup: { type: addressSchema, required: true },
  dropoff: { type: addressSchema, required: true },
  recipient: { type: recipientSchema },
  package: { weight: { type: Number, required: true, min: 0 }, dimensions: { type: dimensionsSchema }, description: { type: String, trim: true } },
  stage: { type: String, enum: ['pending', 'assigned', 'picked_up', 'in_transit', 'delivered', 'failed'], default: 'pending', index: true },
  proofOfDeliveryUrl: { type: String, trim: true },
  priceBreakdown: { type: priceBreakdownSchema, required: true }
}, { timestamps: true, versionKey: false });

orderSchema.index({ 'pickup.location': '2dsphere' });
orderSchema.index({ 'dropoff.location': '2dsphere' });
export type Order = InferSchemaType<typeof orderSchema>;
export const OrderModel = model<Order>('Order', orderSchema);

File: mobile/lib/core/models/order.dart
enum OrderStage { pending, assigned, pickedUp, inTransit, delivered, failed }

extension OrderStageLabel on OrderStage {
  String get label => switch (this) {
    OrderStage.pending => 'Pending',
    OrderStage.assigned => 'Assigned',
    OrderStage.pickedUp => 'Picked up',
    OrderStage.inTransit => 'In transit',
    OrderStage.delivered => 'Delivered',
    OrderStage.failed => 'Failed',
  };

  String get apiValue => name == 'pickedUp'
      ? 'picked_up'
      : name == 'inTransit'
      ? 'in_transit'
      : name;
}

class StopAddress {
  const StopAddress({
    required this.addressLine,
    required this.city,
    this.region,
    this.postalCode,
    this.coordinates = const [],
  });
  final String addressLine;
  final String city;
  final String? region;
  final String? postalCode;
  final List<double> coordinates;

  factory StopAddress.fromJson(Map<String, dynamic> json) {
    final location = json['location'] as Map<String, dynamic>?;
    return StopAddress(
      addressLine: json['addressLine'] as String? ?? 'Address unavailable',
      city: json['city'] as String? ?? '',
      region: json['region'] as String?,
      postalCode: json['postalCode'] as String?,
      coordinates: (location?['coordinates'] as List? ?? [])
          .map((value) => (value as num).toDouble())
          .toList(),
    );
  }
}

class DeliveryOrder {
  const DeliveryOrder({
    required this.id,
    required this.trackingNumber,
    required this.stage,
    required this.pickup,
    required this.dropoff,
    required this.weight,
    this.description,
    this.total = 0,
    this.currency = 'USD',
  });
  final String id;
  final String trackingNumber;
  final OrderStage stage;
  final StopAddress pickup;
  final StopAddress dropoff;
  final double weight;
  final String? description;
  final double total;
  final String currency;

  factory DeliveryOrder.fromJson(Map<String, dynamic> json) {
    final breakdown = json['priceBreakdown'] as Map<String, dynamic>? ?? {};
    return DeliveryOrder(
      id: json['_id'] as String? ?? json['id'] as String? ?? '',
      trackingNumber: json['trackingNumber'] as String? ?? 'Unknown parcel',
      stage: OrderStage.values.firstWhere(
        (item) => item.apiValue == json['stage'],
        orElse: () => OrderStage.assigned,
      ),
      pickup: StopAddress.fromJson(
        (json['pickup'] as Map<String, dynamic>?) ?? const {},
      ),
      dropoff: StopAddress.fromJson(
        (json['dropoff'] as Map<String, dynamic>?) ?? const {},
      ),
      weight:
          ((json['package'] as Map<String, dynamic>?)?['weight'] as num?)
              ?.toDouble() ??
          0,
      description:
          (json['package'] as Map<String, dynamic>?)?['description'] as String?,
      total: (breakdown['total'] as num?)?.toDouble() ?? 0,
      currency: breakdown['currency'] as String? ?? 'USD',
    );
  }
}
